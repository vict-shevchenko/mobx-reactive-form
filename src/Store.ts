import { observable } from 'mobx';
import { Form } from './Form';
import { IFormDefinition } from './interfaces/Form';

export class FormStore {
	@observable public forms: Map<string, Form> = observable.map();

	public registerForm(
			name: string,
			options: IFormDefinition): Form
		{ // tslint:disable-line
		let form: Form;

		if (this.hasForm(name)) {
			form = this.getForm(name) as Form;
			form.extendConfiguration(options);
		}
		else {
			form = new Form(options);
			this.forms.set(name, form);
		}

		return form;
	}

	public unRegisterForm(name: string): void {
		if (this.forms.has(name)) {
			const form = this.getForm(name) as Form;
			form.cleanup();
			this.forms.delete(name);
		}
	}

	public hasForm(name: string): boolean {
		return this.forms.has(name);
	}

	public getForm(name: string): Form | undefined {
		return this.forms.get(name);
	}
}
