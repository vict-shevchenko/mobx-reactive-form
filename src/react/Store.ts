import { observable } from 'mobx';
import { Form } from '../core/Form';
import { IFormDefinition, IFormSchema } from '../interfaces/Form';
import { submitCallback } from '../types';

export class FormStore {
	@observable public forms: Map<string, Form<any>> = observable.map();

	public registerForm<V>(name: string, submit: submitCallback<V>, options: IFormDefinition) {
		let form: Form<V>;

		if (!this.hasForm(name)) {
			form = new Form<V>(submit, options);
			this.forms.set(name, form);
		} else {
			form = this.getForm(name) as Form<V>;
			if (form.attached) {
				// attempt to double register form
				throw new Error(`Form with name "${name}" already exist. Use "withFormData" HOC to extend it`);
			}
		}

		// looks ok, we created new form or retrieved form form store as it was disconnected with keepState=true
		form.attached = true;
		return form;
	}

	public extendForm(name: string, options: { schema: IFormSchema }) {
		if (this.hasForm(name)) {
			const form = this.getForm(name) as Form;
			form.extendConfiguration(options);
			return form;
		} else {
			// tslint:disable-next-line:max-line-length
			throw new Error(
				`Form with name "${name}" does not exist so can not be extended. Use "reactiveMobxForm" HOC to create it firstly`
			);
		}
	}

	public unRegisterForm(name: string) {
		const form = this.getForm(name) as Form;

		form.cleanup();
		this.forms.delete(name);
	}

	public detachForm(name) {
		const form = this.getForm(name) as Form;
		form.attached = false;
	}

	public hasForm(name: string) {
		return this.forms.has(name);
	}

	public getForm(name: string) {
		return this.forms.get(name);
	}
}
