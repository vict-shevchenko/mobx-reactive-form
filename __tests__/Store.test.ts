import { FormStore } from '../lib/Store';

describe('Testing the FormStore', () => {
	let formStore: FormStore;

	beforeAll(() => {
		formStore = new FormStore();
	});

	test('Creation on new FormStore instance should wokr correctly', () => {
		expect(formStore).toEqual(expect.any(FormStore));
	});

	test('created instance of FormStore should have proper methods and properties', () => {
		expect(formStore.forms).toBeDefined();
		expect(formStore.forms.size).toBe(0);

		/* formStore.registerForm('myform', {}, {}, {});

		expect(formStore.forms.get('myform')).toBeDefined();

		formStore.unRegisterForm('myform');

		expect(formStore.forms.get('myform')).not.toBeDefined();
		expect(formStore.forms.size).toBe(0); */
	});

	test('It should be possible to register new Form', () => {
		formStore.registerForm('myform', {});
		expect(formStore.forms.size).toBe(1);
	});

	test('It should not be possible to register new Form with same name', () => {
		formStore.registerForm('myform', {});
		expect(formStore.forms.size).toBe(1);

		formStore.registerForm('myform', {});
		expect(formStore.forms.size).toBe(1);
		expect(formStore.forms.get('myform')!.formSchema).toEqual({});
	});

	test('It should be possible to extend form configuration', () => {
		formStore.registerForm('myform', {});
		expect(formStore.forms.size).toBe(1);
		expect(formStore.forms.get('myform')!.formSchema).toEqual({});

		formStore.registerForm('myform', { schema: { name: ['Viktor', 'required'] } });
		expect(formStore.forms.size).toBe(1);
		expect(formStore.forms.get('myform')!.formSchema).toEqual({ name: ['Viktor', 'required'] });
	});

	test('It should be possible to unregister form', () => {
		formStore.registerForm('myform', {});
		expect(formStore.forms.size).toBe(1);

		formStore.unRegisterForm('myform');
		expect(formStore.forms.size).toBe(0);
	});

	test('It should be possible. Unregistring of form with invalid name should not throw', () => {
		formStore.registerForm('myform', {});
		expect(formStore.forms.size).toBe(1);

		formStore.unRegisterForm('yourform');
		expect(formStore.forms.size).toBe(1);
	});

	test('It should be possible to check form existance', () => {
		formStore.registerForm('myform', {});
		expect(formStore.hasForm('myform')).toBe(true);
		expect(formStore.hasForm('yourform')).toBe(false);
	});

	test('It should be possible to fetch form', () => {
		formStore.registerForm('myform', { schema: { name: ['Viktor', 'required'] } });
		const form = formStore.getForm('myform');
		const yourForm = formStore.getForm('yourform');

		expect(form).toBeDefined();
		expect(form!.formSchema).toEqual({ name: ['Viktor', 'required'] });

		expect(yourForm).not.toBeDefined();
	});
});
