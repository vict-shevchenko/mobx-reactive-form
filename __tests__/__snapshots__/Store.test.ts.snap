// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the FormStore It should not be possible to register new Form with same name, should throw 1`] = `"Form with name \\"myform\\" already exist. Use \\"withFormData\\" HOC to extend it"`;

exports[`Testing the FormStore It should throw when extending form that was not created 1`] = `"Form with name \\"myform\\" does not exist so can not be extended. Use \\"reactiveMobxForm\\" HOC to create it firstly"`;
